client/dashboard/sites/site/environment-switcher.tsx

```diff
@@ -0,0 +1,195 @@
+		</HStack>
+	);
+};
+
+const CurrentEnvironment = ( { site }: { site: Site } ) => {
+	if ( site.is_wpcom_staging_site ) {
+		return <Environment env="staging" />;
+	}
+
+	return <Environment env="production" />;
+};
+
+const EnvironmentSwitcherDropdown = ( {
+	currentSite,
+	otherEnvironment,
+	otherEnvironmentSite,
+	onClose,
+}: {
+	currentSite: Site;
+	otherEnvironment: Env;
+	otherEnvironmentSite?: Site;
+	onClose: () => void;
+} ) => {
+	const productionSite = otherEnvironment === 'staging' ? currentSite : otherEnvironmentSite;
+	const stagingSite = otherEnvironment === 'staging' ? otherEnvironmentSite : currentSite;
+	const { createSuccessNotice, createErrorNotice } = useDispatch( noticesStore );
+	const mutation = useMutation( stagingSiteCreateMutation( productionSite?.ID ?? 0 ) );
+	const handleCreate = () => {
+		mutation.mutate( undefined, {
+			onSuccess: () => {
+				createSuccessNotice( __( 'Staging site created.' ), { type: 'snackbar' } );
+			},
+			onError: ( error: Error ) => {
+				createErrorNotice(
+					sprintf(
+						// translators: "reason" is why adding the staging site failed.
+						__( 'Failed to create staging site: %(reason)s' ),
+						{ reason: error.message }
+					),
+					{
+						type: 'snackbar',
+					}
+				);
+			},
+		} );
+	};
+
+	// TODO: Handle upsell.
+	const handleUpsell = () => {};
+
+	return (
+		<MenuGroup>
+			{ productionSite && canManageSite( productionSite ) && (
+				<RouterLinkMenuItem to={ `/sites/${ productionSite.slug }` } onClick={ onClose }>
+					<Environment env="production" />
+				</RouterLinkMenuItem>
+			) }
+			{ stagingSite && canManageSite( stagingSite ) && (
+				<RouterLinkMenuItem to={ `/sites/${ stagingSite.slug }` } onClick={ onClose }>
+					<Environment env="staging" />
+				</RouterLinkMenuItem>
+			) }
+			{ otherEnvironment === 'staging' && productionSite && ! stagingSite && (
+				<MenuItem onClick={ canCreateStagingSite( productionSite ) ? handleCreate : handleUpsell }>
+					<HStack justify="flex-start">
+						{ mutation.isPending ? (
+							<>
+								<Spinner style={ { width: '24px', height: '24px', padding: '4px', margin: 0 } } />
+								<span>{ __( 'Creating staging site…' ) }</span>
+							</>
+						) : (
+							<>
+								<Icon icon={ plus } />
+								<span>{ __( 'Add staging site' ) }</span>
+							</>
+						) }
+					</HStack>
+				</MenuItem>
+			) }
+		</MenuGroup>
+	);
+};
+
+const EnvironmentSwitcher = ( { site }: { site: Site } ) => {
+	const hasStagingSite =
+		! site.is_wpcom_staging_site && site?.options?.wpcom_staging_blog_ids?.length;
+
+	const otherEnvironment = site.is_wpcom_staging_site ? 'production' : 'staging';
+	const otherEnvironmentSiteId = site.is_wpcom_staging_site
+		? site.options?.wpcom_production_blog_id
+		: site.options?.wpcom_staging_blog_ids?.[ 0 ];
+
+	const { data: otherEnvironmentSite } = useQuery( {
+		...siteByIdQuery( otherEnvironmentSiteId ?? 0 ),
+		enabled: !! otherEnvironmentSiteId,
+	} );
+
+	return (
```

Comment:
Maybe as a follow-up, we should not render the switcher for self-hosted JP sites.

Reply:
Done by https://github.com/Automattic/wp-calypso/pull/104710/commits/874e361a579595556a72cffba1ecc2488715ae29.